src/invariant/ChangedAccountsSubentriesCountIsValid.cpp

```diff
@@ -1,148 +0,0 @@
-{
-    return app.getInvariantManager()
-        .registerInvariant<ChangedAccountsSubentriesCountIsValid>(
-            app.getDatabase());
-}
-
-AccountID
-getAccount(LedgerEntry const& entry)
-{
-    auto& d = entry.data;
-    switch (d.type())
-    {
-    case ACCOUNT:
-        return d.account().accountID;
-    case TRUSTLINE:
-        return d.trustLine().accountID;
-    case OFFER:
-        return d.offer().sellerID;
-    case DATA:
-        return d.data().accountID;
-    default:
-        abort();
-    }
-}
-
-AccountID
-getAccount(LedgerKey const& key)
-{
-    switch (key.type())
-    {
-    case ACCOUNT:
-        return key.account().accountID;
-    case TRUSTLINE:
-        return key.trustLine().accountID;
-    case OFFER:
-        return key.offer().sellerID;
-    case DATA:
-        return key.data().accountID;
-    default:
-        abort();
-    }
-}
-
-std::set<AccountID>
-getAddedOrUpdatedAccounts(LedgerDelta const& delta)
-{
-    auto result = std::set<AccountID>{};
-    for (auto const& c : delta.getChanges())
-    {
-        switch (c.type())
-        {
-        case LEDGER_ENTRY_CREATED:
-            result.insert(getAccount(c.created()));
-            break;
-        case LEDGER_ENTRY_UPDATED:
-            result.insert(getAccount(c.updated()));
-            break;
-        case LEDGER_ENTRY_REMOVED:
-            if (c.removed().type() != ACCOUNT)
-            {
-                result.insert(getAccount(c.removed()));
-            }
-            break;
-        default:
-            break;
-        }
-    }
-    return result;
-}
-
-std::set<AccountID>
-getDeletedAccounts(LedgerDelta const& delta)
-{
-    auto result = std::set<AccountID>{};
-    for (auto const& c : delta.getChanges())
-    {
-        if (c.type() == LEDGER_ENTRY_REMOVED && c.removed().type() == ACCOUNT)
-        {
-            result.insert(getAccount(c.removed()));
-        }
-    }
-    return result;
-}
-
-ChangedAccountsSubentriesCountIsValid::ChangedAccountsSubentriesCountIsValid(
-    Database& db)
-    : mDb{db}
-{
-}
-
-std::string
-ChangedAccountsSubentriesCountIsValid::getName() const
-{
-    return "ChangedAccountsSubentriesCountIsValid";
-}
-
-std::string
-ChangedAccountsSubentriesCountIsValid::checkOnLedgerClose(
-    LedgerDelta const& delta)
-{
-    for (auto const& account : getAddedOrUpdatedAccounts(delta))
-    {
-        auto subentries = numberOfSubentries(account, mDb);
-        if (subentries.inAccountsTable != subentries.calculated)
-        {
-            return fmt::format("account {} subentries count mismatch: "
-                               "{} in accounts table vs {} calculated",
-                               KeyUtils::toStrKey(account),
-                               subentries.inAccountsTable,
-                               subentries.calculated);
-        }
-    }
-
-    for (auto const& account : getDeletedAccounts(delta))
-    {
-        auto subentries = numberOfSubentries(account, mDb);
```

Comment:
NumberOfSubentries seems to not be used, can be deleted now

But maybe we should keep invariant-by-db checking code and call it at checkpoint boundaries?

Reply:
I mean checking content of DB to check if invariants are met. Your new code only check ledger delta. I think it could be useful to check from time to time that DB content is correct.

Reply:
I believe that this is handled by the `CacheIsConsistentWithDatabase` invariant, which was changed in this PR to be checked in `checkOnOperationApply`. The `CacheIsConsistentWithDatabase` invariant checks that the contents of the `LedgerDelta` are also reflected in the database (but not the converse). 

It is possible that there are other changes in the database (for example, made by an external process) that would not be caught by this method. If this is what you want to check for, something that @MonsieurNicolas and I discussed, then this will require more changes in how `CacheIsConsistentWithDatabase` works. But one of the primary design goals for the invariants was to make as few of them access the database as possible, so any additional database checking will likely be added in `CacheIsConsistentWithDatabase`. Let me know if this is what you meant, or if you had something else in mind.